File: code/sdk/android/cloudsheet-Android-sdk/src/main/kotlin/com/opencloudsheet/factory/OneDriveFactory.kt
package com.opencloudsheet.factory

import android.app.Activity
import android.util.Log
import com.opencloudsheet.Provider
import com.opencloudsheet.auth.IAuthenticator
import com.opencloudsheet.auth.OneDriveAuthenticator
import com.opencloudsheet.storage.folder.create.ICreateFolder
import com.opencloudsheet.storage.folder.create.OneDrivePersonalFolderCreator
import com.opencloudsheet.storage.folder.list.IListDirectoryContents
import com.opencloudsheet.storage.folder.list.OneDriveListDirectory
import com.opencloudsheet.model.file.ICloudFile
import com.opencloudsheet.model.workbook.IWorkBook
import com.opencloudsheet.model.workbook.OneDriveWorkBook
import com.opencloudsheet.storage.workbook.create.ICreateWorkBook
import com.opencloudsheet.storage.workbook.create.OneDrivePersonalWorkbookCreator
import com.opencloudsheet.storage.workbook.delete.IDeleteWorkBook
import com.opencloudsheet.storage.workbook.delete.OneDrivePersonalWorkbookDeleter
import com.opencloudsheet.config.OneDriveConfiguration
import com.opencloudsheet.metadata.MetadataManager
import com.opencloudsheet.metadata.OneDriveWorkBookMetadataInfo
import com.opencloudsheet.model.worksheet.IWorksheetRow
import com.opencloudsheet.constants.OneDriveClient
import com.opencloudsheet.interceptor.OneDriveRequestInterceptor
import com.opencloudsheet.utilities.OneDriveResponseHelper
import com.opencloudsheet.utilities.OneDriveWorkbookSessionHelper

/**
 * Microsoft OneDrive implementation of cloud storage factory.
 *
 * Provides Microsoft OneDrive-specific implementations:
 * - OneDriveListDirectory - List directory contents
 * - OneDrivePersonalFolderCreator - Create folders
 * - OneDrivePersonalWorkbookCreator - Create Excel workbooks
 * - OneDriveWorkBook - Workbook operations
 *
 * Dependency Graph (DAG - Directed Acyclic Graph):
 * ```
 * Authenticator (lazy created)
 *     ↓
 * SessionHelper (lazy, depends on Authenticator)
 *     ↓
 * RequestInterceptor (lazy, depends on SessionHelper)
 *     ↓
 * OneDriveClient (lazy, depends on RequestInterceptor)
 *     ↓
 * ListDirectory, CreateFolder, etc. (lazy, depends on Authenticator + Client)
 *     ↓
 * BaseCloudStorageFactory.initializeMetadataManager() (uses CreateFolder + CreateWorkBook)
 * ```
 */
internal class OneDriveFactory(
    private val config: OneDriveConfiguration,
    appName: String
) : BaseCloudStorageFactory(appName) {

    companion object {
        private const val TAG = "OneDriveFactory"
    }

    private val _authenticator: IAuthenticator by lazy {
        OneDriveAuthenticator(config)
    }

    private val _sessionHelper: OneDriveWorkbookSessionHelper by lazy {
        OneDriveWorkbookSessionHelper(_authenticator)
    }

    private val _requestInterceptor: OneDriveRequestInterceptor by lazy {
        OneDriveRequestInterceptor(_sessionHelper)
    }

    private val _oneDriveClient: OneDriveClient by lazy {
        OneDriveClient(_requestInterceptor)
    }

    private val _listDirectory: IListDirectoryContents by lazy {
        OneDriveListDirectory(_authenticator, _oneDriveClient)
    }

    private val _createFolder: ICreateFolder by lazy {
        OneDrivePersonalFolderCreator(_authenticator, getListDirectory(), _oneDriveClient)
    }

    private val _createWorkBook: ICreateWorkBook by lazy {
        OneDrivePersonalWorkbookCreator(_authenticator, _oneDriveClient)
    }

    private val _deleteWorkBook: IDeleteWorkBook by lazy {
        OneDrivePersonalWorkbookDeleter(_authenticator, _oneDriveClient)
    }

    suspend fun initialize(activity: Activity) {
        if (isInitialized) {
            Log.d(TAG, "Already initialized, skipping")
            return
        }

        try {
            Log.d(TAG, "Starting Microsoft OneDrive factory initialization")
            _authenticator.login(activity)
            Log.d(TAG, "User logged in successfully. Token: ${_authenticator.getAuthToken()}")
            initializeMetadataManager()
            Log.d(TAG, "Microsoft OneDrive factory initialized successfully")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize Microsoft OneDrive factory", e)
            throw e
        }
    }

    fun getListDirectory(): IListDirectoryContents = _listDirectory
    override fun getCreateFolder(): ICreateFolder = _createFolder

    override fun getCreateWorkBook(): ICreateWorkBook = _createWorkBook

    override fun getAuthenticator(): IAuthenticator = _authenticator

    override fun getListDirectoryContents(): IListDirectoryContents = _listDirectory

    override fun getDeleteWorkBook(): IDeleteWorkBook = _deleteWorkBook

    override fun <T : IWorksheetRow> createWorkBookInstance(
        clazz: Class<T>,
        workBookEntry: MetadataManager.WorkBookEntry
    ): IWorkBook<T> {
        return OneDriveWorkBook(
            oneDriveWorkBookMetadataInfo(workBookEntry.providerMetadataInfo),
            clazz,
            _authenticator,
            workBookEntry,
            _oneDriveClient
        )
    }

    override suspend fun <T> createMetadataFileWorkbookEntry(
        file: ICloudFile,
        clazz: Class<T>
    ): MetadataManager.WorkBookEntry {
        val ownerId = _authenticator.getUserDetails()?.id()
        val providerMetadataInfo = OneDriveWorkBookMetadataInfo(
            ownerId = ownerId ?: "me",
            fileId = file.getId()
        )
        return MetadataManager.WorkBookEntry(
            name = file.getName(),
            provider = Provider.OneDrive.name,
            description = "This is for the Metadata Information of the Users WorkBooks",
            providerMetadataInfo = OneDriveResponseHelper.toString(providerMetadataInfo)
        )
    }

    override suspend fun getProviderMetadataInfo(
        workbookFile: ICloudFile,
        iosClassName: String?,
        androidClassName: String
    ): String {
        val userId: String? = _authenticator.getUserDetails()?.id()
        return OneDriveResponseHelper.toString(
            OneDriveWorkBookMetadataInfo(
                ownerId = userId!!,
                fileId = workbookFile.getId(),
                iosClassName = iosClassName,
                androidClassName = androidClassName
            )
        )
    }

    override suspend fun deleteWorkBookFile(workbookEntry: MetadataManager.WorkBookEntry) {
        val metadataInfo = OneDriveResponseHelper.fromJson(
            workbookEntry.providerMetadataInfo,
            OneDriveWorkBookMetadataInfo::class.java
        )
        _deleteWorkBook.deleteWorkBook(metadataInfo.ownerId, metadataInfo.fileId)
        Log.d(TAG, "Deleted OneDrive file: ${workbookEntry.name}")
    }

    private fun oneDriveWorkBookMetadataInfo(providerMetadataInfo: String): OneDriveWorkBookMetadataInfo {
        return OneDriveResponseHelper.fromJson(
            providerMetadataInfo,
            OneDriveWorkBookMetadataInfo::class.java
        )
    }

}

File: code/sdk/ios/OpenCloudSheet/Sources/OpenCloudSheet/Factory/OneDriveFactory.swift
//
//  OneDriveFactory.swift
//  OpenCloudSheet
//
//  OneDrive implementation of cloud storage factory
//

import Foundation

final class OneDriveFactory: BaseCloudStorageFactory, @unchecked Sendable {
    let appName: String
    private let config: OneDriveConfiguration

    private var isInitialized: Bool = false
    private var _metadataManager: MetadataManager?
    private var _dataFolder: ICloudFile?

    private lazy var _authenticator: IAuthenticator = {
        return OneDriveAuthenticator(config: config)
    }()

    private lazy var _sessionHelper: OneDriveWorkbookSessionHelper = {
        return OneDriveWorkbookSessionHelper(authenticator: _authenticator)
    }()

    private lazy var _oneDriveClient: OneDriveClient = {
        return OneDriveClient(sessionHelper: _sessionHelper)
    }()

    private lazy var _listDirectory: IListDirectoryContents = {
        return OneDriveListDirectory(authProvider: _authenticator, oneDriveClient: _oneDriveClient)
    }()

    private lazy var _createFolder: ICreateFolder = {
        return OneDrivePersonalFolderCreator(
            authProvider: _authenticator,
            listDirectoryContents: _listDirectory,
            oneDriveClient: _oneDriveClient
        )
    }()

    private lazy var _createWorkBook: ICreateWorkBook = {
        return OneDrivePersonalWorkbookCreator(authProvider: _authenticator, oneDriveClient: _oneDriveClient)
    }()

    private lazy var _deleteWorkBook: IDeleteWorkBook = {
        return OneDrivePersonalWorkbookDeleter(authenticator: _authenticator, oneDriveClient: _oneDriveClient)
    }()

    init(config: OneDriveConfiguration, appName: String) {
        self.config = config
        self.appName = appName
    }

    func initialize(from viewController: PlatformViewController) async throws {
        if isInitialized {
            print("Already initialized, skipping")
            return
        }

        do {
            print("Starting Microsoft OneDrive factory initialization")
            try await _authenticator.login(from: viewController)
            if let token = try await _authenticator.getAuthToken() {
                print("User logged in successfully. Token: \(token)")
            }
            try await initializeMetadataManager()
            print("Microsoft OneDrive factory initialized successfully")
        } catch {
            print("Failed to initialize Microsoft OneDrive factory: \(error)")
            throw error
        }
    }

    func getAuthenticator() -> IAuthenticator {
        return _authenticator
    }

    func getCreateFolder() -> ICreateFolder {
        return _createFolder
    }

    func getCreateWorkBook() -> ICreateWorkBook {
        return _createWorkBook
    }

    func getDeleteWorkBook() -> IDeleteWorkBook {
        return _deleteWorkBook
    }

    func getListDirectoryContents() -> IListDirectoryContents {
        return _listDirectory
    }

    func getMetadataManager() throws -> MetadataManager {
        guard let manager = _metadataManager else {
            throw NSError(domain: "OneDriveFactory", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "Factory not initialized. Call initialize() first."])
        }
        return manager
    }

    func getDataFolder() throws -> ICloudFile {
        guard let folder = _dataFolder else {
            throw NSError(domain: "OneDriveFactory", code: -2,
                         userInfo: [NSLocalizedDescriptionKey: "Factory not initialized. Call initialize() first."])
        }
        return folder
    }

    func createWorkBookInstance(workBookEntry: WorkBookEntry) throws -> (any IWorkBook)? {
        let metadataInfo = oneDriveWorkBookMetadataInfo(providerMetadataInfo: workBookEntry.providerMetadataInfo)
        return OneDriveWorkBook(
            metadataInfo: metadataInfo,
            authenticator: _authenticator,
            workBookEntry: workBookEntry,
            oneDriveClient: _oneDriveClient
        )
    }

    func createMetadataFileWorkbookEntry(file: ICloudFile) async throws -> WorkBookEntry {
        let ownerId = try await _authenticator.getUserDetails()?.id() ?? "me"
        let providerMetadataInfo = OneDriveWorkBookMetadataInfo(
            ownerId: ownerId,
            fileId: file.getId()
        )

        let encoder = JSONEncoder()
        let data = try encoder.encode(providerMetadataInfo)
        let jsonString = String(data: data, encoding: .utf8) ?? "{}"

        return WorkBookEntry(
            name: file.getName(),
            provider: Provider.OneDrive.rawValue,
            description: "This is for the Metadata Information of the Users WorkBooks",
            providerMetadataInfo: jsonString
        )
    }

    func getProviderMetadataInfo(workbookFile: ICloudFile) async throws -> String {
        let userId = try await _authenticator.getUserDetails()?.id() ?? "me"
        let metadataInfo = OneDriveWorkBookMetadataInfo(
            ownerId: userId,
            fileId: workbookFile.getId()
        )

        let encoder = JSONEncoder()
        let data = try encoder.encode(metadataInfo)
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    func deleteWorkBook(workbookEntry: WorkBookEntry) async throws {
        try await deleteWorkBookFile(workbookEntry: workbookEntry)
        guard let manager = _metadataManager else {
            throw NSError(domain: "OneDriveFactory", code: -3,
                         userInfo: [NSLocalizedDescriptionKey: "Factory not initialized"])
        }
        try await manager.deleteWorkBook(workbookEntry: workbookEntry)
    }

    func deleteWorkBookFile(workbookEntry: WorkBookEntry) async throws {
        let metadataInfo = oneDriveWorkBookMetadataInfo(providerMetadataInfo: workbookEntry.providerMetadataInfo)
        try await _deleteWorkBook.deleteWorkBook(ownerId: metadataInfo.ownerId, fileId: metadataInfo.fileId)
        print("Deleted OneDrive file: \(workbookEntry.name)")
    }

    private func initializeMetadataManager() async throws {
        if isInitialized {
            print("Already initialized, skipping")
            return
        }

        print("Starting metadata manager initialization for app: \(appName)")
        let cloudsheetFolder = try await getOrCreateFolder(parent: nil, name: "cloudsheet")
        print("Got or created cloudsheet folder: \(cloudsheetFolder.getId())")
        let appFolder = try await getOrCreateFolder(parent: cloudsheetFolder, name: appName)
        print("Got or created app folder: \(appFolder.getId())")
        _dataFolder = try await getOrCreateFolder(parent: appFolder, name: "data")
        print("Got or created data folder: \(_dataFolder!.getId())")
        let metadataFile = try await getOrCreateWorkbook(parentDirectory: _dataFolder, name: "Metadata")
        print("Got or created Metadata workbook: \(metadataFile.getId())")

        guard let metadataWorkbook = try createWorkBookInstance(
            workBookEntry: try await createMetadataFileWorkbookEntry(file: metadataFile)
        ) else {
            throw NSError(domain: "OneDriveFactory", code: -4,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to create metadata workbook instance"])
        }
        try await metadataWorkbook.initialize()

        _metadataManager = MetadataManager(
            workbook: metadataWorkbook,
            createWorkBookInstance: { [weak self] entry in
                return try self?.createWorkBookInstance(workBookEntry: entry)
            }
        )

        try await _metadataManager!.initialize()
        isInitialized = true
        print("Metadata manager initialization completed successfully")
    }

    private func getOrCreateFolder(parent: ICloudFile?, name: String) async throws -> ICloudFile {
        let contents = try await _listDirectory.listDirectoryContents(directory: parent)
        if let existingFolder = contents.first(where: { $0.isFolder() && $0.getName() == name }) {
            return existingFolder
        }
        return try await _createFolder.createFolder(parentDirectory: parent, folderName: name)
    }

    private func getOrCreateWorkbook(parentDirectory: ICloudFile?, name: String) async throws -> ICloudFile {
        let contents = try await _listDirectory.listDirectoryContents(directory: parentDirectory)
        let fileName = name.hasSuffix(".xlsx") ? name : "\(name).xlsx"
        if let existingFile = contents.first(where: { !$0.isFolder() && $0.getName() == fileName }) {
            return existingFile
        }
        return try await _createWorkBook.createWorkbook(folder: parentDirectory, name: name)
    }

    private func oneDriveWorkBookMetadataInfo(providerMetadataInfo: String) -> OneDriveWorkBookMetadataInfo {
        guard let data = providerMetadataInfo.data(using: .utf8),
              let metadataInfo = try? JSONDecoder().decode(OneDriveWorkBookMetadataInfo.self, from: data) else {
            fatalError("Failed to parse OneDriveWorkBookMetadataInfo from JSON")
        }
        return metadataInfo
    }
}
